### PR TITLE
BLUE-315 Add support for s3 Requester Pays

### PR DESCRIPTION
**Pull request by narendra-altiscale**
Changes into the 2.7.7 branch for support of S3 Requester Pays feature.

### CODE REVIEW

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java [hunk 1]
```diff
@@ -37,19 +37,12 @@
...
-import com.amazonaws.services.s3.model.CannedAccessControlList;
-import com.amazonaws.services.s3.model.DeleteObjectRequest;
-import com.amazonaws.services.s3.model.DeleteObjectsRequest;
-import com.amazonaws.services.s3.model.ListObjectsRequest;
-import com.amazonaws.services.s3.model.ObjectListing;
-import com.amazonaws.services.s3.model.ObjectMetadata;
-import com.amazonaws.services.s3.model.PutObjectRequest;
-import com.amazonaws.services.s3.model.CopyObjectRequest;
-import com.amazonaws.services.s3.model.S3ObjectSummary;
+import com.amazonaws.services.s3.model.*;
```
Comment (jack-altiscale) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java]: Don't use the wildcard importing. Only import exactly what is needed

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java [hunk 2]
```diff
@@ -55,8 +60,8 @@ public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Cli
     this.pos = 0;
     this.closed = false;
     this.wrappedStream = null;
+    this.isRequesterPays = isRequesterPays;
   }
-
```
Comment (jack-altiscale) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java]: Let's not remove whitespace when it isn't necessary. Yes, Hadoop has a whitespace issue but it isn't the responsibility of this patch to fix it

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java [hunk 3]
```diff
@@ -87,7 +92,7 @@ private synchronized void reopen(long pos) throws IOException {
 
     GetObjectRequest request = new GetObjectRequest(bucket, key);
     request.setRange(pos, contentLength-1);
-
```
Comment (jack-altiscale) [inline hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java]: Ditto

**Review by jack-altiscale — CHANGES_REQUESTED** [review]
Needs some work @narendra-altiscale . In addition to the comments I provided, there are over 30 checkstyle issues this patch introduces. Use the maven checkstyle:checkstyle goal to identify the lines you introduced. We don't need to fix existing ones, just the lines you are adding/modifying

File: hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java (hunk 3 above)
Comment (narendra-altiscale) [reply hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java]: @jack-altiscale 
1. Fixed the import * issue. Turns out my intellij setting is set to do a import * when it exceeds 10 classes.
2. Fixed all the check style issues related to this patch.
3. Added the extra spaces that were part of the earlier source code.
Please check..

Thanks,
Narendra

**Comment by jack-altiscale** [comment]
Great. Looks much better, thanks.

I think it would also be a good idea if we got a unit test or two for the new behavior

**Comment by jack-altiscale** [comment]
@narendra-altiscale I ran test-patch on your changes and there still appears to be a few checkstyle issues. I have posted the report and checkstyle issues below. Let's get those resolved, add a test if possible, and I'll merge it.

Test patch report:
```
-1 overall

| Vote |      Subsystem |  Runtime   | Comment
============================================================================
|   0  |        reexec  |   0m 15s   | Docker mode activated. 
|      |                |            | Prechecks 
|  -1  |    test4tests  |   0m  0s   | The patch doesn't appear to include any 
|      |                |            | new or modified tests. Please justify why
|      |                |            | no new tests are needed for this patch.
|      |                |            | Also please list what manual steps were
|      |                |            | performed to verify this patch.
|      |                |            | branch-2.7.7-alti-rebase Compile Tests 
|  +1  |    mvninstall  |   4m 51s   | branch-2.7.7-alti-rebase passed 
|  +1  |    checkstyle  |   0m 14s   | branch-2.7.7-alti-rebase passed 
|      |                |            | Patch Compile Tests 
|  +1  |    mvninstall  |   0m 11s   | the patch passed 
|  -1  |    checkstyle  |   0m 11s   | hadoop-tools/hadoop-aws: The patch 
|      |                |            | generated 21 new + 341 unchanged - 1
|      |                |            | fixed = 362 total (was 342)
|      |                |            | Other Tests 
|  -1  |          unit  |   0m 11s   | hadoop-tools/hadoop-aws: The patch 
|      |                |            | generated 21 new + 341 unchanged - 1
|      |                |            | fixed = 362 total (was 342)
|      |                |   5m 57s   | 


|| Subsystem || Report/Notes ||
============================================================================
| Docker | Client=18.06.1-ce Server=18.06.1-ce Image:yetus/hadoop:9011fbc |
| Optional Tests |  dupname  compile  mvninstall  unit  shadedclient  checkstyle  |
| uname | Linux 4b047e83d395 4.15.0-39-generic #42-Ubuntu SMP Tue Oct 23 15:48:01 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | /testptch/patchprocess/precommit/personality/provided.sh |
| git revision | branch-2.7.7-alti-rebase / a20ba90ab5 |
| maven | version: Apache Maven 3.3.9 (bb52d8502b132ec0a5a3f4c09453c07478323dc5; 2015-11-10T16:41:47+00:00) |
| checkstyle | http://localhost:8080/job/Precommit/6/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
| unit | http://localhost:8080/job/Precommit/6/artifact/out/diff-checkstyle-hadoop-tools_hadoop-aws.txt |
|  Test Results | http://localhost:8080/job/Precommit/6/testReport/ |
| Max. process+thread count | 42 (vs. ulimit of 7000) |
| modules | C: hadoop-tools/hadoop-aws U: hadoop-tools/hadoop-aws |
| Console output | http://localhost:8080/job/Precommit/6/console |
| Powered by | Apache Yetus 0.8.0   http://yetus.apache.org |
```

Checkstyle diff:
```
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AFileSystem.java:98:  /**: First sentence should end with a period.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:47:  /**: First sentence should end with a period.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:54:    this(bucket,key,contentLength,client,stats,false);:17: ',' is not followed by whitespace.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:54:    this(bucket,key,contentLength,client,stats,false);:21: ',' is not followed by whitespace.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:54:    this(bucket,key,contentLength,client,stats,false);:35: ',' is not followed by whitespace.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:54:    this(bucket,key,contentLength,client,stats,false);:42: ',' is not followed by whitespace.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:54:    this(bucket,key,contentLength,client,stats,false);:48: ',' is not followed by whitespace.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,: Line is longer than 80 characters (found 93).
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:3: Missing a Javadoc comment.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:25: Parameter bucket should be final.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:32: 'bucket' hides a field.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:40: Parameter key should be final.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:47: 'key' hides a field.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:52: Parameter contentLength should be final.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:57: 'contentLength' hides a field.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:72: Parameter client should be final.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:56:  public S3AInputStream(String bucket, String key, long contentLength, AmazonS3Client client,:87: 'client' hides a field.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:57:                        FileSystem.Statistics stats, boolean isRequesterPays) {:25: Parameter stats should be final.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:57:                        FileSystem.Statistics stats, boolean isRequesterPays) {:47: 'stats' hides a field.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:57:                        FileSystem.Statistics stats, boolean isRequesterPays) {:54: Parameter isRequesterPays should be final.
./hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/S3AInputStream.java:57:                        FileSystem.Statistics stats, boolean isRequesterPays) {:62: 'isRequesterPays' hides a field.
```

**Comment by narendra-altiscale** [comment]
@jack-altiscale 
I fixed these . can you please check? 
Thanks,
Narendra

**Review by jack-altiscale — APPROVED** [review]

**Comment by jack-altiscale** [comment]
Thanks, @narendra-altiscale. I reviewed your document that contained your reasons for lack of testing. 

LGTM

**Pull request merged**